atomisticparsers/h5md/parser.py

```diff
@@ -0,0 +1,787 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import os
+import numpy as np
+import logging
+import h5py
+
+# from nomad.units import ureg
+from nomad.parsing.file_parser import FileParser
+from nomad.datamodel.metainfo.simulation.run import Run, Program  # TimeRun
+from nomad.datamodel.metainfo.simulation.method import (
+    Method, ForceField, Model, AtomParameters, ForceCalculations, NeighborSearching
+)
+from nomad.datamodel.metainfo.simulation.system import (
+    AtomsGroup
+)
+from nomad.datamodel.metainfo.simulation.calculation import (
+    Calculation, Energy, EnergyEntry, BaseCalculation
+)
+from nomad.datamodel.metainfo.workflow import (
+    Workflow
+)
+from nomad.datamodel.metainfo.simulation.workflow import (
+    EnsembleProperty, EnsemblePropertyValues, CorrelationFunction, CorrelationFunctionValues
+)
+from atomisticparsers.utils import MDParser
+from .metainfo.h5md import ParamEntry, CalcEntry, Author
+from nomad.units import ureg
+
+MOL = 6.022140857e+23
+
+
+class HDF5Parser(FileParser):
+    def __init__(self):
+        super().__init__(None)
+
+    @property
+    def filehdf5(self):
+        if self._file_handler is None:
+            try:
+                self._file_handler = h5py.File(self.mainfile, 'r')
+            except Exception:
+                self.logger.error('Error reading hdf5 file.')
+
+        return self._file_handler
+
+    def apply_unit(self, quantity, unit, unit_factor):
+
+        if quantity is None:
+            return
+        if unit:
+            unit = ureg(unit)
+            unit *= unit_factor
+            quantity *= unit
+
+        return quantity
+
+    def decode_hdf5_bytes(self, dataset, default):
+        if dataset is None:
+            return
+        elif type(dataset).__name__ == 'ndarray':
+            if dataset.size == 0:
+                return default
+            dataset = [val.decode("utf-8") for val in dataset] if isinstance(dataset[0], bytes) else dataset
+            dataset = [val.__bool__() for val in dataset] if type(dataset[0]).__name__ == 'bool_' else dataset
+        elif type(dataset).__name__ == 'bool_':
+            dataset = dataset.__bool__()
+        else:
+            dataset = dataset.decode("utf-8") if isinstance(dataset, bytes) else dataset
+        return dataset
+
+    def hdf5_attr_getter(self, source, attribute, path=None, default=None):
+        '''
+        Extracts attribute from object based on path, and returns default if not defined.
+        '''
+        if path:
+            section_segments = path.split('.')
+            for section in section_segments:
+                try:
+                    value = source.get(section)
+                    source = value[-1] if isinstance(value, list) else value
+                except Exception:
+                    return
+        value = source.attrs.get(attribute)
+        source = value[-1] if isinstance(value, list) else value
```

Comment:
maybe use value or another variable here instead of source.

Reply:
sure. done.